File: examples/typescript/authed_websocket.ts
// Authenticating with an API key over the trading WebSocket.
//
// After the socket opens, the first frame must be a signed ApiKeySignIn (mt: 29).
// See authed_rest_requests.ts for how the API key is loaded from the environment.
import * as ed from '@noble/ed25519';
import { randomBytes } from 'crypto';
import { fileURLToPath } from 'url';
import WebSocket from 'ws';

import { loadApiKey } from './authed_rest_requests.js';


const CHAIN_ID = Number(process.env.PERPL_CHAIN_ID) || 143;

const PERPL_WS_URL = process.env.PERPL_WS_URL || 'wss://app.perpl.xyz';
const TRADING_API = '/ws/v1/trading';

// MsgTypeApiKeySignIn: the API-key WebSocket sign-in frame type.
const API_KEY_SIGN_IN = 29;


interface ApiKeySignInMessage {
    mt: number;
    chain_id: number;
    api_key: string;
    timestamp: string;
    nonce: string;
    signature: string;
}


async function signIn(ws: WebSocket, chainId: number, token: string, privateKey: Uint8Array): Promise<void> {
    const timestamp = Date.now().toString();
    const nonce = randomBytes(16).toString('base64url');

    const canonical = [chainId, 'trading-ws-signin', timestamp, nonce].join('\n');
    const signature = Buffer.from(await ed.signAsync(Buffer.from(canonical), privateKey)).toString('base64url');

    const message: ApiKeySignInMessage = {
        mt: API_KEY_SIGN_IN,
        chain_id: chainId,
        api_key: token,
        timestamp,
        nonce,
        signature,
    };
    ws.send(JSON.stringify(message));

    ws.on('message', (data: WebSocket.RawData) => {
        console.log(`Received: ${data}`);
    });
}

function authedWebsocket(wsUrl: string, chainId: number, token: string, privateKey: Uint8Array): void {
    const url = wsUrl + TRADING_API;
    const ws = new WebSocket(url);

    ws.on('open', () => {
        signIn(ws, chainId, token, privateKey);
    });

    ws.on('error', (err: Error) => {
        console.error('WebSocket error:', err);
    });
}

async function main(): Promise<void> {
    const { token, privateKey } = loadApiKey();
    authedWebsocket(PERPL_WS_URL, CHAIN_ID, token, privateKey);
}


if (process.argv[1] === fileURLToPath(import.meta.url)) {
    await main();
}

File: examples/python/authed_websocket.py
# Authenticating the trading WebSocket with an API key you already have.
#
# On the trading socket, API-key auth is a signed ApiKeySignIn frame (mt: 29)
# sent as the first message after the socket opens. Key handling and the signing
# helpers are shared with the REST example.
#
# Create a key via the web UI (connect your wallet):
#   Mainnet: https://app.perpl.xyz/apikeys
#   Testnet: https://testnet.perpl.xyz/apikeys
# Then export PERPL_API_KEY and PERPL_API_KEY_SECRET (see authed_rest_requests.py).
import asyncio
import base64
import json
import os
import secrets
import time

import websockets

from authed_rest_requests import (
    API_URL,
    PERPL_CHAIN_ID,
    load_api_key,
)


PERPL_WS_URL = os.environ.get("PERPL_WS_URL", "wss://app.perpl.xyz")
TRADING_API = "/ws/v1/trading"
API_KEY_SIGN_IN = 29  # MsgTypeApiKeySignIn


def _b64url(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode()


async def start_stream(ws, priv, token, chain_id):
    # First frame: signed ApiKeySignIn. The signature covers the WS canonical
    # string: chain_id \n "trading-ws-signin" \n timestamp_ms \n nonce.
    timestamp = str(int(time.time() * 1000))
    nonce = _b64url(secrets.token_bytes(16))
    canonical = "\n".join([str(chain_id), "trading-ws-signin", timestamp, nonce])
    sig = _b64url(priv.sign(canonical.encode()))

    sign_in = {
        "mt": API_KEY_SIGN_IN,
        "chain_id": chain_id,
        "api_key": token,
        "timestamp": timestamp,
        "nonce": nonce,
        "signature": sig,
    }
    await ws.send(json.dumps(sign_in))

    async for message in ws:
        print(f"Received: {message}")


async def authed_websocket(ws_url, priv, token, chain_id):
    url = ws_url + TRADING_API
    async with websockets.connect(url) as ws:
        await start_stream(ws, priv, token, chain_id)


async def main():
    token, priv = load_api_key()
    await authed_websocket(PERPL_WS_URL, priv, token, PERPL_CHAIN_ID)


if __name__ == "__main__":
    asyncio.run(main())
